In [5]:
import pandas as pd
import numpy as np

In [48]:
Trans=pd.read_csv(r"C:\Users\gaura\Downloads\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey CSV\transaction_data.csv")
Trans

,household_key,BASKET_ID,DAY,PRODUCT_ID,QUANTITY,SALES_VALUE,STORE_ID,RETAIL_DISC,TRANS_TIME,WEEK_NO,COUPON_DISC,COUPON_MATCH_DISC
0,2375,26984851472,1,1004906,1,1.39,364,-0.60,1631,1,0.0,0.0
1,2375,26984851472,1,1033142,1,0.82,364,0.00,1631,1,0.0,0.0
2,2375,26984851472,1,1036325,1,0.99,364,-0.30,1631,1,0.0,0.0
3,2375,26984851472,1,1082185,1,1.21,364,0.00,1631,1,0.0,0.0
4,2375,26984851472,1,8160430,1,1.50,364,-0.39,1631,1,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...
2595727,1598,42305362535,711,92130,1,0.99,3228,0.00,1520,102,0.0,0.0
2595728,1598,42305362535,711,114102,1,8.89,3228,0.00,1520,102,0.0,0.0
2595729,1598,42305362535,711,133449,1,6.99,3228,0.00,1520,102,0.0,0.0
2595730,1598,42305362535,711,6923644,1,4.50,3228,-0.49,1520,102,0.0,0.0


In [49]:
Trans.size

31148784

In [50]:
Trans.shape

(2595732, 12)

In [51]:
Trans.dtypes

household_key          int64
BASKET_ID              int64
DAY                    int64
PRODUCT_ID             int64
QUANTITY               int64
SALES_VALUE          float64
STORE_ID               int64
RETAIL_DISC          float64
TRANS_TIME             int64
WEEK_NO                int64
COUPON_DISC          float64
COUPON_MATCH_DISC    float64
dtype: object

In [52]:
Trans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2595732 entries, 0 to 2595731
Data columns (total 12 columns):
 #   Column             Dtype  
---  ------             -----  
 0   household_key      int64  
 1   BASKET_ID          int64  
 2   DAY                int64  
 3   PRODUCT_ID         int64  
 4   QUANTITY           int64  
 5   SALES_VALUE        float64
 6   STORE_ID           int64  
 7   RETAIL_DISC        float64
 8   TRANS_TIME         int64  
 9   WEEK_NO            int64  
 10  COUPON_DISC        float64
 11  COUPON_MATCH_DISC  float64
dtypes: float64(4), int64(8)
memory usage: 237.6 MB


In [53]:
Trans.head()

,household_key,BASKET_ID,DAY,PRODUCT_ID,QUANTITY,SALES_VALUE,STORE_ID,RETAIL_DISC,TRANS_TIME,WEEK_NO,COUPON_DISC,COUPON_MATCH_DISC
0,2375,26984851472,1,1004906,1,1.39,364,-0.60,1631,1,0.0,0.0
1,2375,26984851472,1,1033142,1,0.82,364,0.00,1631,1,0.0,0.0
2,2375,26984851472,1,1036325,1,0.99,364,-0.30,1631,1,0.0,0.0
3,2375,26984851472,1,1082185,1,1.21,364,0.00,1631,1,0.0,0.0
4,2375,26984851472,1,8160430,1,1.50,364,-0.39,1631,1,0.0,0.0


In [54]:
Trans.tail()

,household_key,BASKET_ID,DAY,PRODUCT_ID,QUANTITY,SALES_VALUE,STORE_ID,RETAIL_DISC,TRANS_TIME,WEEK_NO,COUPON_DISC,COUPON_MATCH_DISC
2595727,1598,42305362535,711,92130,1,0.99,3228,0.00,1520,102,0.0,0.0
2595728,1598,42305362535,711,114102,1,8.89,3228,0.00,1520,102,0.0,0.0
2595729,1598,42305362535,711,133449,1,6.99,3228,0.00,1520,102,0.0,0.0
2595730,1598,42305362535,711,6923644,1,4.50,3228,-0.49,1520,102,0.0,0.0
2595731,1598,42305362535,711,14055192,1,6.99,3228,0.00,1520,102,0.0,0.0


In [55]:
Trans.isnull().sum()

household_key        0
BASKET_ID            0
DAY                  0
PRODUCT_ID           0
QUANTITY             0
SALES_VALUE          0
STORE_ID             0
RETAIL_DISC          0
TRANS_TIME           0
WEEK_NO              0
COUPON_DISC          0
COUPON_MATCH_DISC    0
dtype: int64

In [56]:
Trans.duplicated().sum()

np.int64(0)

In [57]:
Trans.describe()

,household_key,BASKET_ID,DAY,PRODUCT_ID,QUANTITY,SALES_VALUE,STORE_ID,RETAIL_DISC,TRANS_TIME,WEEK_NO,COUPON_DISC,COUPON_MATCH_DISC
count,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06,2.595732e+06
mean,1.271953e+03,3.402620e+10,3.887562e+02,2.891435e+06,1.004286e+02,3.104120e+00,3.142673e+03,-5.387054e-01,1.561586e+03,5.622150e+01,-1.641600e-02,-2.918564e-03
std,7.260660e+02,4.711649e+09,1.897210e+02,3.837404e+06,1.153436e+03,4.182274e+00,8.937113e+03,1.249191e+00,3.998378e+02,2.710223e+01,2.168410e-01,3.969004e-02
min,1.000000e+00,2.698485e+10,1.000000e+00,2.567100e+04,0.000000e+00,0.000000e+00,1.000000e+00,-1.800000e+02,0.000000e+00,1.000000e+00,-5.593000e+01,-7.700000e+00
25%,6.560000e+02,3.040805e+10,2.290000e+02,9.174590e+05,1.000000e+00,1.290000e+00,3.300000e+02,-6.900000e-01,1.308000e+03,3.300000e+01,0.000000e+00,0.000000e+00
50%,1.272000e+03,3.276081e+10,3.900000e+02,1.028816e+06,1.000000e+00,2.000000e+00,3.720000e+02,-1.000000e-02,1.613000e+03,5.600000e+01,0.000000e+00,0.000000e+00
75%,1.913000e+03,4.012685e+10,5.530000e+02,1.133018e+06,1.000000e+00,3.490000e+00,4.220000e+02,0.000000e+00,1.843000e+03,8.000000e+01,0.000000e+00,0.000000e+00
max,2.500000e+03,4.230536e+10,7.110000e+02,1.831630e+07,8.963800e+04,8.400000e+02,3.428000e+04,3.990000e+00,2.359000e+03,1.020000e+02,0.000000e+00,0.000000e+00


In [58]:
Trans.nunique()

household_key          2500
BASKET_ID            276484
DAY                     711
PRODUCT_ID            92339
QUANTITY              11746
SALES_VALUE            5502
STORE_ID                582
RETAIL_DISC            2414
TRANS_TIME             1440
WEEK_NO                 102
COUPON_DISC             415
COUPON_MATCH_DISC        80
dtype: int64

In [59]:
(Trans["SALES_VALUE"] < 0).sum()

np.int64(0)

In [60]:
(Trans["RETAIL_DISC"] < 0).sum()

np.int64(1303026)

In [61]:
Trans["QUANTITY"].describe()

count    2.595732e+06
mean     1.004286e+02
std      1.153436e+03
min      0.000000e+00
25%      1.000000e+00
50%      1.000000e+00
75%      1.000000e+00
max      8.963800e+04
Name: QUANTITY, dtype: float64

In [62]:
(Trans["SALES_VALUE"] == 0).sum()

np.int64(18850)

In [63]:
Trans.memory_usage(deep=True).sum()/1024**2

np.float64(237.64648818969727)

In [64]:
Trans[Trans["SALES_VALUE"] == 0].head(20)

,household_key,BASKET_ID,DAY,PRODUCT_ID,QUANTITY,SALES_VALUE,STORE_ID,RETAIL_DISC,TRANS_TIME,WEEK_NO,COUPON_DISC,COUPON_MATCH_DISC
97,744,26985165432,1,5978648,0,0.0,31582,0.00,1119,1,0.00,0.0
128,1287,26985336468,1,5978648,0,0.0,304,0.00,1351,1,0.00,0.0
249,2305,26996870743,2,5978656,0,0.0,414,0.00,1300,1,-1.00,0.0
293,271,26997082949,2,5978656,0,0.0,329,0.00,1719,1,-2.00,0.0
694,315,27008952267,3,957951,0,0.0,327,0.00,1707,1,0.00,0.0
766,2110,27009082349,3,837963,1,0.0,375,0.00,1547,1,0.00,0.0
996,568,27021090189,4,951412,1,0.0,446,-1.48,35,1,0.00,0.0
1089,14,27021203242,4,6463658,0,0.0,319,0.00,1704,1,0.00,0.0
1095,2110,27021248054,4,850115,1,0.0,375,-3.29,1705,1,0.00,0.0
1113,2110,27021248054,4,976872,1,0.0,375,-2.49,1705,1,0.00,0.0


In [65]:
Trans[Trans["SALES_VALUE"] == 0][["SALES_VALUE","RETAIL_DISC","COUPON_DISC","COUPON_MATCH_DISC"]].describe()


,SALES_VALUE,RETAIL_DISC,COUPON_DISC,COUPON_MATCH_DISC
count,18850.0,18850.000000,18850.000000,18850.000000
mean,0.0,-0.781153,-0.550489,-0.000289
std,0.0,1.816268,1.505729,0.011747
min,0.0,-39.950000,-37.930000,-0.500000
25%,0.0,0.000000,-0.500000,0.000000
50%,0.0,0.000000,0.000000,0.000000
75%,0.0,0.000000,0.000000,0.000000
max,0.0,0.010000,0.000000,0.000000


Data Cleaning Decision:
Identified 18,850 transactions with SALES_VALUE = 0. After examining associated retail and coupon discounts, these transactions appear to represent promotional or fully discounted purchases rather than data errors. Therefore, these records were retained for analysis.

In [66]:
Trans.sort_values("QUANTITY", ascending=False).head(20)

,household_key,BASKET_ID,DAY,PRODUCT_ID,QUANTITY,SALES_VALUE,STORE_ID,RETAIL_DISC,TRANS_TIME,WEEK_NO,COUPON_DISC,COUPON_MATCH_DISC
1750942,630,34749153595,503,6534178,89638,250.00,384,-13.45,1927,73,0.0,0.0
468356,2407,29392047893,181,6544236,85055,210.00,375,-8.51,1606,27,0.0,0.0
481876,630,29484790880,185,6534178,61335,150.21,384,-6.13,1056,27,0.0,0.0
166536,149,28210551971,103,6534178,51912,110.00,327,-5.19,1514,15,0.0,0.0
1340882,193,32956767959,402,6534178,48073,121.10,310,-7.21,1825,58,0.0,0.0
2472791,2133,41904760458,682,6534178,45475,100.00,433,-4.55,1025,98,0.0,0.0
1560340,149,33768630428,456,6534178,41833,115.00,327,-4.18,1059,66,0.0,0.0
376686,149,29035716247,158,6534178,41686,100.00,327,-4.17,1310,23,0.0,0.0
156049,1406,28167562655,99,6544236,41485,85.00,327,-4.15,2015,15,0.0,0.0
87625,107,27865225627,74,6534178,39365,72.00,329,-3.94,1031,11,0.0,0.0


In [67]:
Trans["QUANTITY"].value_counts().head(20)

QUANTITY
1     2050967
2      372373
3       66962
4       38667
0       14466
5       10918
6        8981
8        2531
10       2219
7        1379
12        909
9         863
11        208
20        186
15        178
14        136
16        122
13        107
24         97
18         52
Name: count, dtype: int64

In [68]:
(Trans["QUANTITY"] == 0).sum()

np.int64(14466)

In [69]:
pdt=pd.read_csv(r"C:\Users\gaura\Downloads\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey CSV\product.csv")
pdt

,PRODUCT_ID,MANUFACTURER,DEPARTMENT,BRAND,COMMODITY_DESC,SUB_COMMODITY_DESC,CURR_SIZE_OF_PRODUCT
0,25671,2,GROCERY,National,FRZN ICE,ICE - CRUSHED/CUBED,22 LB
1,26081,2,MISC. TRANS.,National,NO COMMODITY DESCRIPTION,NO SUBCOMMODITY DESCRIPTION,
2,26093,69,PASTRY,Private,BREAD,BREAD:ITALIAN/FRENCH,
3,26190,69,GROCERY,Private,FRUIT - SHELF STABLE,APPLE SAUCE,50 OZ
4,26355,69,GROCERY,Private,COOKIES/CONES,SPECIALTY COOKIES,14 OZ
...,...,...,...,...,...,...,...
92348,18293142,6384,DRUG GM,National,BOOKSTORE,PAPERBACK BOOKS,
92349,18293439,6393,DRUG GM,National,BOOKSTORE,CHILDRENS LOW END,
92350,18293696,6406,DRUG GM,National,BOOKSTORE,PAPERBACK BEST SELLER,
92351,18294080,6442,DRUG GM,National,BOOKSTORE,PAPERBACK BOOKS,


In [70]:
pdt.columns

Index(['PRODUCT_ID', 'MANUFACTURER', 'DEPARTMENT', 'BRAND', 'COMMODITY_DESC',
       'SUB_COMMODITY_DESC', 'CURR_SIZE_OF_PRODUCT'],
      dtype='object')

In [71]:
pdt[pdt["PRODUCT_ID"].isin([6534178, 6544236, 6534166])]

,PRODUCT_ID,MANUFACTURER,DEPARTMENT,BRAND,COMMODITY_DESC,SUB_COMMODITY_DESC,CURR_SIZE_OF_PRODUCT
57216,6534166,69,MISC SALES TRAN,Private,COUPON/MISC ITEMS,GASOLINE-REG UNLEADED,
57221,6534178,69,KIOSK-GAS,Private,COUPON/MISC ITEMS,GASOLINE-REG UNLEADED,
57335,6544236,69,MISC SALES TRAN,Private,COUPON/MISC ITEMS,GASOLINE-REG UNLEADED,


In [72]:
pdt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92353 entries, 0 to 92352
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   PRODUCT_ID            92353 non-null  int64 
 1   MANUFACTURER          92353 non-null  int64 
 2   DEPARTMENT            92353 non-null  object
 3   BRAND                 92353 non-null  object
 4   COMMODITY_DESC        92353 non-null  object
 5   SUB_COMMODITY_DESC    92353 non-null  object
 6   CURR_SIZE_OF_PRODUCT  92353 non-null  object
dtypes: int64(2), object(5)
memory usage: 4.9+ MB


In [73]:
pdt.isnull().sum()

PRODUCT_ID              0
MANUFACTURER            0
DEPARTMENT              0
BRAND                   0
COMMODITY_DESC          0
SUB_COMMODITY_DESC      0
CURR_SIZE_OF_PRODUCT    0
dtype: int64

In [74]:
pdt.duplicated().sum()

np.int64(0)

In [75]:
pdt["PRODUCT_ID"].duplicated().sum()

np.int64(0)

In [76]:
pdt.nunique()

PRODUCT_ID              92353
MANUFACTURER             6476
DEPARTMENT                 44
BRAND                       2
COMMODITY_DESC            308
SUB_COMMODITY_DESC       2383
CURR_SIZE_OF_PRODUCT     4345
dtype: int64

In [77]:
pdt.describe(include="all")

,PRODUCT_ID,MANUFACTURER,DEPARTMENT,BRAND,COMMODITY_DESC,SUB_COMMODITY_DESC,CURR_SIZE_OF_PRODUCT
count,9.235300e+04,92353.00000,92353,92353,92353,92353,92353
unique,NaN,NaN,44,2,308,2383,4345
top,NaN,NaN,GROCERY,National,GREETING CARDS/WRAP/PARTY SPLY,CARDS EVERYDAY,
freq,NaN,NaN,39021,78537,2785,1005,30607
mean,5.328353e+06,1739.22833,NaN,NaN,NaN,NaN,NaN
std,5.359937e+06,1818.26957,NaN,NaN,NaN,NaN,NaN
min,2.567100e+04,1.00000,NaN,NaN,NaN,NaN,NaN
25%,9.706280e+05,328.00000,NaN,NaN,NaN,NaN,NaN
50%,1.621091e+06,1094.00000,NaN,NaN,NaN,NaN,NaN
75%,9.704770e+06,2264.00000,NaN,NaN,NaN,NaN,NaN


In [78]:
pdt.head()

,PRODUCT_ID,MANUFACTURER,DEPARTMENT,BRAND,COMMODITY_DESC,SUB_COMMODITY_DESC,CURR_SIZE_OF_PRODUCT
0,25671,2,GROCERY,National,FRZN ICE,ICE - CRUSHED/CUBED,22 LB
1,26081,2,MISC. TRANS.,National,NO COMMODITY DESCRIPTION,NO SUBCOMMODITY DESCRIPTION,
2,26093,69,PASTRY,Private,BREAD,BREAD:ITALIAN/FRENCH,
3,26190,69,GROCERY,Private,FRUIT - SHELF STABLE,APPLE SAUCE,50 OZ
4,26355,69,GROCERY,Private,COOKIES/CONES,SPECIALTY COOKIES,14 OZ


In [79]:
pdt.tail()

,PRODUCT_ID,MANUFACTURER,DEPARTMENT,BRAND,COMMODITY_DESC,SUB_COMMODITY_DESC,CURR_SIZE_OF_PRODUCT
92348,18293142,6384,DRUG GM,National,BOOKSTORE,PAPERBACK BOOKS,
92349,18293439,6393,DRUG GM,National,BOOKSTORE,CHILDRENS LOW END,
92350,18293696,6406,DRUG GM,National,BOOKSTORE,PAPERBACK BEST SELLER,
92351,18294080,6442,DRUG GM,National,BOOKSTORE,PAPERBACK BOOKS,
92352,18316298,764,GROCERY,National,PAPER TOWELS,PAPER TOWELS & HOLDERS,


In [80]:
pdt.shape

(92353, 7)

In [81]:
pdt.size

646471

In [82]:
pdt.dtypes

PRODUCT_ID               int64
MANUFACTURER             int64
DEPARTMENT              object
BRAND                   object
COMMODITY_DESC          object
SUB_COMMODITY_DESC      object
CURR_SIZE_OF_PRODUCT    object
dtype: object

In [83]:
coupon=pd.read_csv(r"C:\Users\gaura\Downloads\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey CSV\coupon.csv")
coupon

,COUPON_UPC,PRODUCT_ID,CAMPAIGN
0,10000089061,27160,4
1,10000089064,27754,9
2,10000089073,28897,12
3,51800009050,28919,28
4,52100000076,28929,25
...,...,...,...
124543,10000085478,13038802,18
124544,10000085479,13039368,18
124545,10000085478,13039553,18
124546,10000085479,13776332,18


In [84]:
coupon.shape

(124548, 3)

In [85]:
coupon.size

373644

In [86]:
coupon.head()

,COUPON_UPC,PRODUCT_ID,CAMPAIGN
0,10000089061,27160,4
1,10000089064,27754,9
2,10000089073,28897,12
3,51800009050,28919,28
4,52100000076,28929,25


In [87]:
coupon.tail()

,COUPON_UPC,PRODUCT_ID,CAMPAIGN
124543,10000085478,13038802,18
124544,10000085479,13039368,18
124545,10000085478,13039553,18
124546,10000085479,13776332,18
124547,10000085477,13213390,18


In [88]:
coupon["COUPON_UPC"].duplicated().sum()

np.int64(123413)

In [89]:
coupon["PRODUCT_ID"].duplicated().sum()

np.int64(80415)

In [90]:
coupon[coupon.duplicated(keep=False)].sort_values(
    ["COUPON_UPC","PRODUCT_ID","CAMPAIGN"])

,COUPON_UPC,PRODUCT_ID,CAMPAIGN
73199,10000085426,930799,13
73748,10000085426,930799,13
62387,10000085427,983002,13
64785,10000085427,983002,13
64507,10000085427,1078612,13
...,...,...,...
47353,58304613024,8019845,30
47354,58304613024,8019845,30
47355,58304613024,8019845,30
47356,58304613024,8019845,30


In [91]:
coupon.groupby(
    ["COUPON_UPC","PRODUCT_ID","CAMPAIGN"]
).size().sort_values(ascending=False).head(20)

COUPON_UPC   PRODUCT_ID  CAMPAIGN
57940011080  921912      27          19
57940011075  921912      27          19
57940018081  6704447     27          17
57940018076  6704447     27          17
54589314175  1024433     27          16
54589314187  1024433     27          16
57940018081  7167488     27          16
57940018076  7167488     27          16
51111131075  1065098     27          15
54100027032  1118641     27          15
51111122275  5565415     27          15
51111131082  1065098     27          15
57940041075  933409      27          15
57940041055  933409      27          15
51111122081  5565415     27          15
51111122275  942300      27          14
54589314175  940099      27          14
54589314187  940099      27          14
51111122081  942300      27          14
             1079365     27          14
dtype: int64

In [92]:
coupon.columns

Index(['COUPON_UPC', 'PRODUCT_ID', 'CAMPAIGN'], dtype='object')

In [93]:
coupon.describe

<bound method NDFrame.describe of          COUPON_UPC  PRODUCT_ID  CAMPAIGN
0       10000089061       27160         4
1       10000089064       27754         9
2       10000089073       28897        12
3       51800009050       28919        28
4       52100000076       28929        25
...             ...         ...       ...
124543  10000085478    13038802        18
124544  10000085479    13039368        18
124545  10000085478    13039553        18
124546  10000085479    13776332        18
124547  10000085477    13213390        18

[124548 rows x 3 columns]>

In [94]:
coupon.isnull().sum()

COUPON_UPC    0
PRODUCT_ID    0
CAMPAIGN      0
dtype: int64

In [95]:
coupon.duplicated().sum()

np.int64(5164)

In [96]:
coupon.drop_duplicates().shape

(119384, 3)

In [97]:
coupon = coupon.drop_duplicates()

Coupon Table Cleaning

Missing values: 0
Exact duplicate rows: 5,164
Action: Removed duplicate records after confirming that all column values were identical.
Reason: Duplicate records can inflate coupon counts and lead to incorrect campaign analysis.

In [98]:
cp=pd.read_csv(r"C:\Users\gaura\Downloads\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey CSV\coupon_redempt.csv")
cp

,household_key,DAY,COUPON_UPC,CAMPAIGN
0,1,421,10000085364,8
1,1,421,51700010076,8
2,1,427,54200000033,8
3,1,597,10000085476,18
4,1,597,54200029176,18
...,...,...,...,...
2313,2496,592,54900050076,18
2314,2496,610,55100000013,18
2315,2500,449,53663200076,8
2316,2500,449,54300031076,8


In [99]:
cp.shape

(2318, 4)

In [100]:
cp.size

9272

In [101]:
cp.duplicated().sum()

np.int64(0)

In [102]:
cp.isnull().sum()

household_key    0
DAY              0
COUPON_UPC       0
CAMPAIGN         0
dtype: int64

In [103]:
cp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2318 entries, 0 to 2317
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   household_key  2318 non-null   int64
 1   DAY            2318 non-null   int64
 2   COUPON_UPC     2318 non-null   int64
 3   CAMPAIGN       2318 non-null   int64
dtypes: int64(4)
memory usage: 72.6 KB


In [104]:
cp.nunique()

household_key    434
DAY              328
COUPON_UPC       556
CAMPAIGN          30
dtype: int64

In [105]:
cp.describe(include="all")

,household_key,DAY,COUPON_UPC,CAMPAIGN
count,2318.000000,2318.000000,2.318000e+03,2318.000000
mean,1302.816652,528.217429,4.123049e+10,15.551337
std,783.002545,100.360985,1.986068e+10,5.716636
min,1.000000,225.000000,1.000009e+10,1.000000
25%,588.000000,458.250000,1.000009e+10,13.000000
50%,1396.500000,532.000000,5.234003e+10,14.000000
75%,2004.000000,605.000000,5.430002e+10,18.000000
max,2500.000000,704.000000,5.897850e+10,30.000000


In [106]:
hh=pd.read_csv(r"C:\Users\gaura\Downloads\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey CSV\hh_demographic.csv")
hh

,classification_1,classification_2,classification_3,HOMEOWNER_DESC,classification_5,classification_4,KID_CATEGORY_DESC,household_key
0,Age Group6,X,Level4,Homeowner,Group5,2,None/Unknown,1
1,Age Group4,X,Level5,Homeowner,Group5,2,None/Unknown,7
2,Age Group2,Y,Level3,Unknown,Group4,3,1,8
3,Age Group2,Y,Level6,Homeowner,Group4,4,2,13
4,Age Group4,Z,Level5,Homeowner,Group3,1,None/Unknown,16
...,...,...,...,...,...,...,...,...
796,Age Group3,Y,Level5,Homeowner,Group5,2,None/Unknown,2494
797,Age Group4,X,Level6,Homeowner,Group6,3,1,2496
798,Age Group4,Y,Level4,Unknown,Group2,1,None/Unknown,2497
799,Age Group2,Y,Level5,Homeowner,Group5,2,None/Unknown,2498


In [107]:
hh.shape

(801, 8)

In [108]:
hh.size

6408

In [109]:
hh.isnull().sum()

classification_1     0
classification_2     0
classification_3     0
HOMEOWNER_DESC       0
classification_5     0
classification_4     0
KID_CATEGORY_DESC    0
household_key        0
dtype: int64

In [110]:
hh.duplicated().sum()

np.int64(0)

In [111]:
hh["household_key"].duplicated().sum()

np.int64(0)

In [112]:
hh.describe(include="all")

,classification_1,classification_2,classification_3,HOMEOWNER_DESC,classification_5,classification_4,KID_CATEGORY_DESC,household_key
count,801,801,801,801,801,801,801,801.000000
unique,6,3,12,5,6,5,4,NaN
top,Age Group4,Y,Level5,Homeowner,Group5,2,None/Unknown,NaN
freq,288,344,192,504,255,318,558,NaN
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1235.176030
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,736.804647
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.000000
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,596.000000
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1218.000000
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1914.000000


In [113]:
hh.nunique()

classification_1       6
classification_2       3
classification_3      12
HOMEOWNER_DESC         5
classification_5       6
classification_4       5
KID_CATEGORY_DESC      4
household_key        801
dtype: int64

In [114]:
hh.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 801 entries, 0 to 800
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   classification_1   801 non-null    object
 1   classification_2   801 non-null    object
 2   classification_3   801 non-null    object
 3   HOMEOWNER_DESC     801 non-null    object
 4   classification_5   801 non-null    object
 5   classification_4   801 non-null    object
 6   KID_CATEGORY_DESC  801 non-null    object
 7   household_key      801 non-null    int64 
dtypes: int64(1), object(7)
memory usage: 50.2+ KB


In [115]:
cd=pd.read_csv(r"C:\Users\gaura\Downloads\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey CSV\causal_data.csv")
cd

,PRODUCT_ID,STORE_ID,WEEK_NO,display,mailer
0,26190,286,70,0,A
1,26190,288,70,0,A
2,26190,289,70,0,A
3,26190,292,70,0,A
4,26190,293,70,0,A
...,...,...,...,...,...
36786519,18203921,439,100,1,0
36786520,18203921,439,101,1,0
36786521,18243471,415,101,3,0
36786522,18243476,415,101,3,0


In [116]:
cd.shape

(36786524, 5)

In [117]:
cd.size

183932620

In [118]:
cd.head()

,PRODUCT_ID,STORE_ID,WEEK_NO,display,mailer
0,26190,286,70,0,A
1,26190,288,70,0,A
2,26190,289,70,0,A
3,26190,292,70,0,A
4,26190,293,70,0,A


In [119]:
cd.tail()

,PRODUCT_ID,STORE_ID,WEEK_NO,display,mailer
36786519,18203921,439,100,1,0
36786520,18203921,439,101,1,0
36786521,18243471,415,101,3,0
36786522,18243476,415,101,3,0
36786523,18244453,345,101,7,0


In [120]:
cd.isnull().sum()

PRODUCT_ID    0
STORE_ID      0
WEEK_NO       0
display       0
mailer        0
dtype: int64

In [121]:
cd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36786524 entries, 0 to 36786523
Data columns (total 5 columns):
 #   Column      Dtype 
---  ------      ----- 
 0   PRODUCT_ID  int64 
 1   STORE_ID    int64 
 2   WEEK_NO     int64 
 3   display     object
 4   mailer      object
dtypes: int64(3), object(2)
memory usage: 1.4+ GB


In [122]:
cd.duplicated().sum()

np.int64(0)

In [123]:
cd.nunique()

PRODUCT_ID    68377
STORE_ID        115
WEEK_NO          93
display          10
mailer           11
dtype: int64

In [124]:
cam=pd.read_csv(r"C:\Users\gaura\Downloads\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey CSV\campaign_desc.csv")
cam

,DESCRIPTION,CAMPAIGN,START_DAY,END_DAY
0,TypeB,24,659,719
1,TypeC,15,547,708
2,TypeB,25,659,691
3,TypeC,20,615,685
4,TypeB,23,646,684
5,TypeB,21,624,656
6,TypeB,22,624,656
7,TypeA,18,587,642
8,TypeB,19,603,635
9,TypeB,17,575,607


In [125]:
cam.shape

(30, 4)

In [126]:
cam.size

120

In [127]:
cam.head()

,DESCRIPTION,CAMPAIGN,START_DAY,END_DAY
0,TypeB,24,659,719
1,TypeC,15,547,708
2,TypeB,25,659,691
3,TypeC,20,615,685
4,TypeB,23,646,684


In [128]:
cam.tail()

,DESCRIPTION,CAMPAIGN,START_DAY,END_DAY
25,TypeA,30,323,369
26,TypeB,29,281,334
27,TypeB,28,259,320
28,TypeC,27,237,300
29,TypeA,26,224,264


In [129]:
cam.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   DESCRIPTION  30 non-null     object
 1   CAMPAIGN     30 non-null     int64 
 2   START_DAY    30 non-null     int64 
 3   END_DAY      30 non-null     int64 
dtypes: int64(3), object(1)
memory usage: 1.1+ KB


In [130]:
cam.isnull().sum()

DESCRIPTION    0
CAMPAIGN       0
START_DAY      0
END_DAY        0
dtype: int64

In [131]:
cam.duplicated().sum()

np.int64(0)

In [132]:
cam.nunique()

DESCRIPTION     3
CAMPAIGN       30
START_DAY      27
END_DAY        28
dtype: int64

In [133]:
ct=pd.read_csv(r"C:\Users\gaura\Downloads\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey\dunnhumby_The-Complete-Journey CSV\campaign_table.csv")
ct

,DESCRIPTION,household_key,CAMPAIGN
0,TypeA,17,26
1,TypeA,27,26
2,TypeA,212,26
3,TypeA,208,26
4,TypeA,192,26
...,...,...,...
7203,TypeC,1803,15
7204,TypeC,1082,15
7205,TypeC,942,15
7206,TypeC,855,15


In [134]:
ct.shape

(7208, 3)

In [135]:
ct.size

21624

In [136]:
ct.head()

,DESCRIPTION,household_key,CAMPAIGN
0,TypeA,17,26
1,TypeA,27,26
2,TypeA,212,26
3,TypeA,208,26
4,TypeA,192,26


In [137]:
ct.tail()

,DESCRIPTION,household_key,CAMPAIGN
7203,TypeC,1803,15
7204,TypeC,1082,15
7205,TypeC,942,15
7206,TypeC,855,15
7207,TypeC,334,15


In [138]:
ct.isnull().sum()

DESCRIPTION      0
household_key    0
CAMPAIGN         0
dtype: int64

In [139]:
ct.duplicated().sum()

np.int64(0)

In [140]:
ct.nunique()

DESCRIPTION         3
household_key    1584
CAMPAIGN           30
dtype: int64

In [141]:
ct.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7208 entries, 0 to 7207
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   DESCRIPTION    7208 non-null   object
 1   household_key  7208 non-null   int64 
 2   CAMPAIGN       7208 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 169.1+ KB


In [143]:
# Folder path
save_path = r"C:\Gaurav Personal\DA & DS"

# Save all cleaned tables
Trans.to_csv(f"{save_path}\clean_transaction_data.csv", index=False)

pdt.to_csv(f"{save_path}\clean_product.csv", index=False)

coupon.to_csv(f"{save_path}\clean_coupon.csv", index=False)

cp.to_csv(f"{save_path}\clean_coupon_redempt.csv", index=False)

hh.to_csv(f"{save_path}\clean_hh_demographic.csv", index=False)

cd.to_csv(f"{save_path}\clean_causal_data.csv", index=False)

cam.to_csv(f"{save_path}\clean_campaign_desc.csv", index=False)

ct.to_csv(f"{save_path}\clean_campaign_table.csv", index=False)

print("All 8 cleaned CSV files have been saved successfully!")

✅ All 8 cleaned CSV files have been saved successfully!
